### PR TITLE
Fix issue where inspect and diagnostics didn't include local elastic-agent configuration.

### LINKED ISSUE #2390: proxy_url info not displayed in inspect or diagnostics command
- Version: 8.6.1 (probably back to 8.6.0)
- Operating System: Linux
- Steps to Reproduce:

1. install elastic-agent with `--proxy-url=http://proxy.addr:port`
2. elastic-agent inspect and elastic-agent diagnostics don't display the 'proxy_url' settings even though they are saved and used for connectivity

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Previously when you ran `inspect` for an elastic-agent that was enrolled in Fleet it only provided the configuration that came from the latest policy and would not include the local configuration that the Elastic Agent also used.

Previously when you ran `diagnostics` it would not include the local configuration that the elastic-agent is using it would only provide the configuration from the latest policy.

This fixes both issues stated above.

## Why is it important?

Without this change not enough information was provided from either command to get a true representation of what the elastic-agent is running with.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- ~~[ ] I have added an integration test or an E2E test~~

## Related issues

- Closes #2390

### CODE REVIEW

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 1]
```diff
@@ -424,6 +427,22 @@ func (c *Coordinator) Run(ctx context.Context) error {
 // information about the state of the Elastic Agent.
 func (c *Coordinator) DiagnosticHooks() diagnostics.Hooks {
 	return diagnostics.Hooks{
+		{
+			Name:        "local-config",
+			Filename:    "local-config.yaml",
```
Comment (cmacknz) [inline internal/pkg/agent/application/coordinator/coordinator.go:432]: Let's document this file in the official docs https://github.com/elastic/ingest-docs/blob/main/docs/en/ingest-management/commands.asciidoc#elastic-agent-diagnostics-command

**Review by michalpristas — APPROVED** [review]
change looks good,

**Comment by pierrehilbert** [comment]
@michalpristas: did you test it locally in addition?

**Comment by ycombinator** [comment]
I tested this locally with the use case described in this PR's corresponding issue: https://github.com/elastic/elastic-agent/issues/2390.  Unfortunately, I'm not seeing the proxy URL supplied with `./elastic-agent install --proxy-url` in the output of `elastic-agent inspect`.

```
$ sudo ./elastic-agent install --proxy-url=http://localhost:8888
$ sudo elastic-agent version
Binary: 8.8.0-SNAPSHOT (build: 69b91f82d6b590aee36300e8a5c336ac20f59d85 at 2023-04-25 22:37:58 +0000 UTC)
Daemon: 8.8.0-SNAPSHOT (build: 69b91f82d6b590aee36300e8a5c336ac20f59d85 at 2023-04-25 22:37:58 +0000 UTC)
$ sudo elastic-agent inspect
agent:
  logging:
    to_stderr: true
inputs:
- data_stream.namespace: default
  id: unique-system-metrics-input
  streams:
  - data_stream.dataset: system.cpu
    metricsets:
    - cpu
  - data_stream.dataset: system.memory
    metricsets:
    - memory
  - data_stream.dataset: system.network
    metricsets:
    - network
  - data_stream.dataset: system.filesystem
    metricsets:
    - filesystem
  type: system/metrics
  use_output: default
outputs:
  default:
    api_key: example-key
    hosts:
    - 127.0.0.1:9200
    type: elasticsearch
```

**Review by ycombinator — CHANGES_REQUESTED** [review]
I tested this locally with the _Steps to Reproduce_ mentioned in this PR's corresponding issue: https://github.com/elastic/elastic-agent/issues/2390.  Unfortunately, I'm not seeing the proxy URL supplied with `./elastic-agent install --proxy-url` in the output of `elastic-agent inspect`.

```
$ sudo ./elastic-agent install --proxy-url=http://localhost:8888
$ sudo elastic-agent version
Binary: 8.8.0-SNAPSHOT (build: 69b91f82d6b590aee36300e8a5c336ac20f59d85 at 2023-04-25 22:37:58 +0000 UTC)
Daemon: 8.8.0-SNAPSHOT (build: 69b91f82d6b590aee36300e8a5c336ac20f59d85 at 2023-04-25 22:37:58 +0000 UTC)
$ sudo elastic-agent inspect
agent:
  logging:
    to_stderr: true
inputs:
- data_stream.namespace: default
  id: unique-system-metrics-input
  streams:
  - data_stream.dataset: system.cpu
    metricsets:
    - cpu
  - data_stream.dataset: system.memory
    metricsets:
    - memory
  - data_stream.dataset: system.network
    metricsets:
    - network
  - data_stream.dataset: system.filesystem
    metricsets:
    - filesystem
  type: system/metrics
  use_output: default
outputs:
  default:
    api_key: example-key
    hosts:
    - 127.0.0.1:9200
    type: elasticsearch
```

I'm also not seeing the proxy URL in the `local-config.yaml` file in the diagnostics bundle.

```
$ sudo elastic-agent diagnostics && unzip elastic-agent-diagnostics-*.zip -d diag
[WARNING] Could not redact state.yaml due to unmarshalling error: yaml: invalid map key: map[interface {}]interface {}{"unitid":"system/metrics-default", "unittype":1}
Created diagnostics archive "elastic-agent-diagnostics-2023-04-25T22-50-39Z-00.zip"
***** WARNING *****
Created archive may contain plain text credentials.
Ensure that files in archive are redacted before sharing.
*******************
Archive:  elastic-agent-diagnostics-2023-04-25T22-50-39Z-00.zip
  inflating: diag/version.txt
  inflating: diag/goroutine.txt
  ...
   creating: diag/logs/
  inflating: diag/logs/elastic-agent-20230425.ndjson
$ grep -i proxy diag/local-config.yaml | wc -l
       0
```

File: internal/pkg/agent/application/coordinator/coordinator.go [hunk 2]
```diff
@@ -424,6 +427,22 @@ func (c *Coordinator) Run(ctx context.Context) error {
 // information about the state of the Elastic Agent.
 func (c *Coordinator) DiagnosticHooks() diagnostics.Hooks {
 	return diagnostics.Hooks{
+		{
+			Name:        "local-config",
```
Comment (pchila) [inline internal/pkg/agent/application/coordinator/coordinator.go:431]: If we are adding a new diagnostic hook providing a new file, we should probably test that we get the right data for it in `diagnostics_test.go`

File: internal/pkg/config/config.go [hunk 3]
```diff
@@ -155,19 +142,76 @@ func (c *Config) access() *ucfg.Config {
...
+	// add the skipped keys into the map and back into the config
+	for k, v := range skippedKeys {
+		m[k] = v
+	}
+	if len(skippedKeysOrig) > 0 {
+		err := c.access().Merge(skippedKeysOrig, ucfg.ResolveNOOP)
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (pchila) [inline internal/pkg/config/config.go:214]: I am confused by this block: after adding in the resulting map the skipped keys without expanding their values why do we merge it back into config with the `${<name>}` ?
Moreover it seems strange for a method that creates a `map[string]interface{}` out of a Config object to have side effects, is there is a specific reason for that ?

**Review by pchila — CHANGES_REQUESTED** [review]
I would like to see a test for the new coordinator hook (it doesn't need to fit in the big test we already have for a basic fleet policy + system integration, it can be its own test).
Moreover we should add the new diagnostic file and hook in the existing big test (even if the expected result is an empty file)

Aside from that, I have an open question as to why we need the `ToMapStr` method to have side effects

File: internal/pkg/config/config.go (hunk 3 above)
Comment (blakerouse) [reply internal/pkg/config/config.go:214]: That is because the code above removed the entire field from the original `*Config` (aka. it mutated), so it must add them back in or the configuration after this point would be missing those fields.

Personally I hate this code, this really should be added as a core feature in `go-ucfg` which would result in this not having all this hacking stuff in place.

Without this change it is impossible for the `ToMapString()` to work, because once it hits a variable it will unpack it and try to resolve it, breaking it.

File: internal/pkg/agent/application/coordinator/coordinator.go (hunk 2 above)
Comment (blakerouse) [reply internal/pkg/agent/application/coordinator/coordinator.go:431]: I can add a test. I was surprised that the test did not fail. We should add a check in those tests to fail if a file is added or removed, so we can ensure that we catch and guide a developer to modify the test to get coverage of a file.

**Comment by blakerouse** [comment]
> I tested this locally with the _Steps to Reproduce_ mentioned in this PR's corresponding issue: #2390. Unfortunately, I'm not seeing the proxy URL supplied with `./elastic-agent install --proxy-url` in the output of `elastic-agent inspect`.

@ycombinator I don't think you where testing a build that has this PR. I have just validated this PR again, and it is working. I see both the `proxy_url` in the inspect command output and in the diagnostics output.

File: internal/pkg/agent/application/coordinator/coordinator.go (hunk 2 above)
Comment (blakerouse) [reply internal/pkg/agent/application/coordinator/coordinator.go:431]: I have added to the test in the latest commit. I also added the change to the test to ensure that in the future if a new hook is aded the test fails so it must be added to the test.

**Comment by ycombinator** [comment]
> > I tested this locally with the _Steps to Reproduce_ mentioned in this PR's corresponding issue: #2390. Unfortunately, I'm not seeing the proxy URL supplied with `./elastic-agent install --proxy-url` in the output of `elastic-agent inspect`.
> 
> @ycombinator I don't think you where testing a build that has this PR. I have just validated this PR again, and it is working. I see both the `proxy_url` in the inspect command output and in the diagnostics output.

I did.  I included the output of `elastic-agent version` and you can see the commit SHA there belongs to one of the commits in the PR.  Nevertheless, I'll re-test with the latest commit.

**Comment by ycombinator** [comment]
I just re-tested this PR with the same command that's mentioned in the corresponding issue: https://github.com/elastic/elastic-agent/issues/2390.  And I'm still not seeing the proxy URL in either the output of `sudo elastic-agent inspect` or anywhere in the archive produced by `sudo elastic-agent diagnostics`.

```
$ sudo ./elastic-agent install --proxy-url=http://localhost:9999
Elastic Agent will be installed at /Library/Elastic/Agent and will run as a service. Do you want to continue? [Y/n]:y
Do you want to enroll this Agent into Fleet? [Y/n]:n
Elastic Agent has been successfully installed.
```
```
$ sudo elastic-agent version
Binary: 8.8.0-SNAPSHOT (build: f93fc1754dc4a12ac8f6b8f1ca1b7872548d1e3f at 2023-04-27 05:18:43 +0000 UTC)
Daemon: 8.8.0-SNAPSHOT (build: f93fc1754dc4a12ac8f6b8f1ca1b7872548d1e3f at 2023-04-27 05:18:43 +0000 UTC)
```
```
$ sudo elastic-agent inspect | grep proxy | wc -l
```
```
$ sudo elastic-agent diagnostics && unzip elastic-agent-diagnostics-*.zip -d diag && cd diag
...
$ grep -inr 'proxy.*url' * | wc -l
       0
```

@blakerouse Did you use the same steps as above or something different? Maybe I'm doing something wrong in my test steps?

File: internal/pkg/agent/application/coordinator/testdata/simple_config/expected/local-config.yaml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  timeout: 10m0s
+  agent:
+    id: ""
+agent:
+  id: ""
+  download:
+    operatingsystem: ""
+    architecture: ""
+    sourceuri: https://artifacts.elastic.co/downloads/
+    targetdirectory: /Users/blake/Library/Caches/JetBrains/GoLand2023.1/tmp/GoLand/data/elastic-agent-unknow/downloads
```
Comment (pchila) [inline internal/pkg/agent/application/coordinator/testdata/simple_config/expected/local-config.yaml:15]: these paths are not portable, they should be replaced with placeholder in the yaml and the hook actual output should be sanitized before comparison.
Have a look at [this snippet](https://github.com/elastic/elastic-agent/blob/1ec8d1c21bcc8bff6e0dfcb5be26289b230bc243/internal/pkg/agent/application/coordinator/diagnostics_test.go#L354) done for the `*-config.yaml` files

File: internal/pkg/agent/application/coordinator/diagnostics_test.go [hunk 5]
```diff
@@ -488,6 +500,7 @@ func newCoordinatorTestHelper(t *testing.T, agentInfo *info.AgentInfo, specs com
 
 	helper.coordinator = coordinator.New(
 		log,
+		configuration.DefaultConfiguration(),
```
Comment (pchila) [inline internal/pkg/agent/application/coordinator/diagnostics_test.go:503]: It would be better if the configuration was passed as a parameter to make it an explicit part of the test using the helper

**Review by pchila — COMMENTED** [review]
Thanks for the test on the new hook, I think we need a small adjustment on the paths in the expected yaml as those are specific to your environment but the rest looks good

**Comment by blakerouse** [comment]
@ycombinator I see the issue, you need to enroll the agent into Fleet. `--proxy-url` is for enrollment and communication with Fleet only.

**Comment by ycombinator** [comment]
> @ycombinator I see the issue, you need to enroll the agent into Fleet. `--proxy-url` is for enrollment and communication with Fleet only.

Ah okay, so it ignores that setting unless other enrollment settings are specified as well.  Thanks.  I'll add those and re-test.

File: internal/pkg/agent/application/coordinator/testdata/simple_config/expected/local-config.yaml (hunk 4 above)
Comment (blakerouse) [reply internal/pkg/agent/application/coordinator/testdata/simple_config/expected/local-config.yaml]: I used the suggestion below to adjust the configuration to have static paths. That allows the test to run without having to call sanitize for the resulting hook.

File: internal/pkg/agent/application/coordinator/diagnostics_test.go (hunk 5 above)
Comment (blakerouse) [reply internal/pkg/agent/application/coordinator/diagnostics_test.go]: I updated this to use a created configuration, that adjusts the paths.

File: internal/pkg/agent/application/coordinator/diagnostics_test.go [hunk 6]
```diff
@@ -269,11 +270,21 @@ func TestCoordinatorDiagnosticHooks(t *testing.T) {
...
 			}
+			sort.Strings(hooksNames)
 
+			expectedNames := make([]string, 0, len(expectedDiagnosticHooks))
+			for n, _ := range expectedDiagnosticHooks {
+				expectedNames = append(expectedNames, n)
+			}
+			sort.Strings(expectedNames)
+
+			require.Equal(t, expectedNames, hooksNames)
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply internal/pkg/agent/application/coordinator/diagnostics_test.go]: Only sorted to ensure matching worked well, I switch to `ElementsMatch` as order doesn't matter.

**Comment by ycombinator** [comment]
> > @ycombinator I see the issue, you need to enroll the agent into Fleet. `--proxy-url` is for enrollment and communication with Fleet only.
> 
> Ah okay, so it ignores that setting unless other enrollment settings are specified as well. Thanks. I'll add those and re-test.

Re-tested while enrolling the agent and I can see `proxy_url` now in the output of `sudo elastic-agent inspect` and in the archive produced by `sudo elatic-agent diagnostics`.

File: internal/pkg/config/config.go [hunk 7]
```diff
@@ -155,19 +142,76 @@ func (c *Config) access() *ucfg.Config {
...
 // ToMapStr takes the config and transform it into a map[string]interface{}
-func (c *Config) ToMapStr() (map[string]interface{}, error) {
+func (c *Config) ToMapStr(opts ...interface{}) (map[string]interface{}, error) {
+	if len(opts) == 0 {
+		opts = DefaultOptions
+	}
+	ucfgOpts, local, err := getOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ycombinator) [inline internal/pkg/config/config.go:160]: Just a thought - not sure if it makes sense or not: it looks like the constructor, `NewConfigFrom` takes `opts` and defaults them to `DefaultOptions`, just like we're doing here.  And then it also calls `getOptions` just like here.  Would it make sense to define the `Config` type as a struct and persist either `opts` and/or `ucfgOpts` and `local` as fields in there so we can just use them here?

**Comment by blakerouse** [comment]
@ycombinator @pchila If you are both good with the fixes and changes, could I get a +1 from you. I cannot merge until your "Needs fixing" is +1.

File: internal/pkg/config/config.go (hunk 7 above)
Comment (ycombinator) [reply internal/pkg/config/config.go:160]: Fair enough, not for this PR.

**Review by ycombinator — APPROVED** [review]

**Review by pchila — APPROVED** [review]
